baseplate/frameworks/queue_producer/avro_kafka.py

```diff
@@ -0,0 +1,97 @@
+import logging
+
+from typing import Any
+from typing import Dict
+from typing import Optional
+
+from confluent_kafka.avro import AvroProducer
+
+from baseplate import _ExcInfo
+from baseplate import Span
+from baseplate import SpanObserver
+from baseplate.clients import ContextFactory
+from baseplate.lib import config
+
+logger = logging.getLogger(__name__)
+
+
+class AvroProducerContextFactory(ContextFactory):
+    def __init__(self, bootstrap_servers: str, schema_registry: str, acks: int):
+        self.bootstrap_servers = bootstrap_servers
+        self.schema_registry = schema_registry
+        self.acks = acks
+
+    def make_object_for_context(self, name: str, span: Span) -> AvroProducer:
+        avro_producer = KafkaAvroProducer(
+            name,
+            span,
+            {
+                "bootstrap.servers": self.bootstrap_servers,
+                "schema.registry.url": self.schema_registry,
+                "acks": self.acks,
+            },
+        )
+        span.register(AvroProducerSpanObserver(avro_producer))
+        return avro_producer
+
+
+class KafkaAvroProducer:
+    def __init__(self, name: str, span: Span, app_config: Dict[str, Any]):
+        self.name = name
+        self.span = span
+        self.avro_producer = AvroProducer(app_config)
+
+    def produce(self, topic: str, schema_id: int, value: Dict[str, Any]) -> None:
+        """Encode `value` using the `schema_id` and produce the message to Kafka.
+
+        :param topic: topic name
+        :param schema_id: schema id associated with the schema that `value` conforms to. It will
+                          be used to look up the schema from Schema Registry. This prevents callers
+                          from using unregistered schemas
+       :param value: object to serialize
+
+        """
+        serializer = self.avro_producer._serializer
+
+        encode_trace_name = "{}.{}".format(self.name, "serializer.encode_record_with_schema_id")
+        encode_span = self.span.make_child(encode_trace_name)
+
+        with encode_span:
+            buffer = serializer.encode_record_with_schema_id(schema_id, value)
+
+        producer_trace_name = "{}.{}".format(self.name, "avro_producer.produce")
+        producer_span = self.span.make_child(producer_trace_name)
+
+        producer = super(AvroProducer, self.avro_producer)  # pylint: disable=bad-super-call;
```

Comment:
Why do we have this `super` call here, some quirk of the kafka library?  Could you add a comment about it?

Reply:
can we replace it with `super(AvroProducer, self.avro_producer).produce(topic=topic, value=buffer)` and comment that we're doing this because we're manually encoding avro because the library might try to re-register the schema for us

Reply:
it seems there was conversation in the Thing PR about having to do it this way to avoid linting errors. https://github.snooguts.net/reddit/reddit-service-thing/pull/796/files#r247408

but I will try it. 
The difference is in being able to call the `encode_record_with_schema_id` instead of `encode_record_with_schema` which causes schemas to be registered.

Reply:
The reason we call the parent class' `produce` is because `AvroProducer`s `produce` method tries to register the schema under `<subject_name>-value` automatically. I think a fix was just pushed in to not have it auto register schemas.

So, we decided to encode the message to Avro ourselves by calling https://github.com/confluentinc/confluent-kafka-python/blob/e575850e4e5d21544d590b76afc7d4f9876a04c9/confluent_kafka/avro/serializer/message_serializer.py#L120 instead that does not auto-register schemas.

The reason it was split into two lines was because I ran into a error cycle between pylint and black that both wanted the code to be a certain way.

Reply:
ok I think the way python wants us to do this is

```
Producer.produce(self.avro_producer, topic=topic, value=value)
```